### ISSUE TITLE
Variable incorrectly marked as Unbound

### ISSUE THREAD

**Issue by albireox** [microsoft/pylance-release#4247]
## Environment data

-   Language Server version: v2023.4.21
-   OS and version: macOS 13.3.1
-   Python version (& distribution if applicable, e.g. Anaconda): pyenv 3.9.9 (virtualenv)

## Code Snippet

```python
def fail():
    raise ValueError()


try:
    a = 1
except:
    fail()

print(2 * a)
```

## Repro Steps

This code produces an error indicating that `a` is unbound or `Literal[1]`. However the variable is not unbound since any error would call `fail()` which would raise and error and the `print` statement wouldn't be reached. If one replaces the call to `fail()` with `raise ValueError()` directly, `a` is correctly marked as `Literal[1]` only and there is no `Unbound` error.

**Comment by erictraut** [comment]
If you want the type analyzer to assume that `fail` will raise an exception on all code paths, you need to provide an explicit return type annotation of `NoReturn`, like this:

```python
from typing import NoReturn

def fail() -> NoReturn:
    raise ValueError()
```

Pyright (the type checker that underlies pylance) previously attempted to infer the `NoReturn` type, but this created undesirable behaviors in some cases. It was also extremely costly at analysis time because it needed to infer return types for every call expression to determine reachability. For these reasons, it no longer infers `NoReturn` return types for purposes of code flow analysis. You need to explicitly annotate your function if `NoReturn` is your intent.

**Comment by albireox** [comment]
Thanks for the quick reply. Indeed, that fixes the issue and it's a cheap and easy option, so I'm happy to just annotate the function that raises the error.

Interestingly, it seems that Pylance already detected that was a `NoReturn` function (if you hover over the call to the function it annotates it as `NoReturn`).

**Comment by erictraut** [comment]
Yes, pyright will infer the return type when it's analyzing a function in your code, but it will not use the `NoReturn` inferred return type when performing code flow analysis at call sites of the function. Doing so would require pyright to infer the return types of all called functions — even those located in libraries that your code calls — to determine code reachability. And this would need to be done recursively for all call chains for every call in your code. You can see how this would be very expensive. If you're interested in understanding the thinking behind this decision, you can read more in [this issue](https://github.com/microsoft/pyright/issues/4787).
